lib/service/dataLink.js

```diff
@@ -0,0 +1,214 @@
+  }
+
+  // Handle a bug in KCS - if query is part of the task, put it in the request
+  if (!task.request.query) {
+    task.request.query = task.query;
+  }
+
+  const serviceObjectToProcess = serviceObject(task.request.serviceObjectName);
+  let dataOp = '';
+  const dataLinkCompletionHandler = kinveyCompletionHandler(task, callback);
+
+  try {
+    task.request.body = JSON.parse(task.request.body);
+  } catch (e) {
+    if (task.request.body && typeof task.request.body !== 'object') {
+      const result = task.response;
+      result.body = kinveyErrors.generateKinveyError('BadRequest', 'Request body is not JSON');
+      result.statusCode = result.body.statusCode;
+      delete result.body.statusCode;
+      return callback(task);
+    }
+  }
+
+  try {
+    task.request.query = JSON.parse(task.request.query);
+  } catch (error) {
+    if (task.request.query && typeof task.request.query !== 'object') {
+      const result = task.response;
+      result.body = kinveyErrors.generateKinveyError('BadRequest', 'Request query is not JSON');
+      result.statusCode = result.body.statusCode;
+      delete result.body.statusCode;
+      return callback(task);
+    }
+  }
+
+  if (task.method === 'POST') {
+    dataOp = 'onInsert';
+  } else if (task.method === 'PUT') {
+    dataOp = 'onUpdate';
+  } else if (task.method === 'GET' && task.endpoint !== '_count') {
+    const taskRequest = _.get(task, 'request', {});
+    if (typeof taskRequest.entityId !== 'undefined' && taskRequest.entityId !== null) {
+      dataOp = 'onGetById';
+    } else if (typeof taskRequest.query !== 'undefined' && taskRequest.query !== null) {
+      dataOp = 'onGetByQuery';
+    } else {
+      dataOp = 'onGetAll';
+    }
+  } else if (task.method === 'GET' && task.endpoint === '_count') {
+    if (typeof task.query !== 'undefined' && task.query !== null) {
+      dataOp = 'onGetCountWithQuery';
+    } else {
+      dataOp = 'onGetCount';
+    }
+  } else if (task.method === 'DELETE') {
+    if (task.request.entityId) {
+      dataOp = 'onDeleteById';
+    } else if (typeof task.query !== 'undefined' && task.query !== null) {
+      dataOp = 'onDeleteByQuery';
+    } else {
+      dataOp = 'onDeleteAll';
+    }
+  } else {
+    const result = task.response;
+    result.body = kinveyErrors.generateKinveyError('BadRequest', 'Cannot determine data operation');
+    result.statusCode = result.body.statusCode;
+    delete result.body.statusCode;
+    return callback(task);
+  }
+
+  const operationHandler = serviceObjectToProcess.resolve(dataOp);
+  operationHandler(task.request, (err, result) => dataLinkCompletionHandler(err, result));
+};
+
+exports.removeServiceObject = (serviceObjectToRemove) => {
+  if (typeof serviceObjectToRemove === 'undefined' || serviceObjectToRemove === null) {
+    throw new Error('Must list ServiceObject name');
+  }
+
+  delete registeredServiceObjects[serviceObjectToRemove];
+};
+
+exports.clearAll = () => {
+  registeredServiceObjects = {};
```

Comment:
This won't work for `const`.  However, if we change to maps like I suggest above, you can use `registeredServiceObjects.clear() here.